### DISCUSSION TITLE
dynamic datagrid updating

### DISCUSSION THREAD

**Discussion by cyriljcb** [AvaloniaUI/Avalonia#14975] [Q&A] [answered]
hello, I have a datagrid refresh problem as the title says. Basically, I'm trying to pass labels contained in datasets between two datagrids. 
datagrid initialization is fine, but once my datagrids are initialized and I just need to update them, no changes are displayed in the GUI. I'm sure my data is fine. How can I solve this problem? If ever, if I rebuild my datagrid completely each time, it works as it should. However, it's not optimized at all. 

Here is my code behind : 

public partial class MainView : UserControl
{
    private MainViewModel _viewModel;
    private bool _isInitializedL = false;
    private bool _isInitializedR = false;
    public MainView()
    {
        InitializeComponent();
        this.Loaded += MainView_Loaded;
    }
    private void MainView_Loaded(object sender, EventArgs e)
    {
        _viewModel = DataContext as MainViewModel;
        if (_viewModel != null)
        {
            _viewModel.ItemsDataSetUpdated += ViewModel_ItemsDataSetUpdated;
            _viewModel.SelectItemsUpdated += ViewModel_SelectItemsUpdated;
        }
    }
    private void ViewModel_ItemsDataSetUpdated(object sender, EventArgs e)
    {
        if( !_isInitializedL)
        {
            InitializeGridL();
            _isInitializedL = true;
        }
        else
        {
            UpdateGridL();

        }
    }

    private void ViewModel_SelectItemsUpdated(object sender, EventArgs e)
    {
        if (!_isInitializedR)
        {
            InitializeGridR();
            _isInitializedR = true;
        }
        else
        {
            UpdateGridR();

        }
    }
    private void InitializeGridL()
    {
        if (_viewModel.ItemsDataSet != null && _viewModel.ItemsDataSet.Tables.Count > 0)
        {
            GridL.Columns.Clear();
            GridL.ItemsSource = _viewModel.ItemsDataSet.Tables[0].DefaultView;

            foreach (DataColumn column in _viewModel.ItemsDataSet.Tables[0].Columns)
            {
                if (column.DataType == typeof(bool))
                {
                    GridL.Columns.Add(new DataGridCheckBoxColumn { Header = column.ColumnName, Binding = new Avalonia.Data.Binding($"Row.ItemArray[{column.Ordinal}]") });
                }
                else
                {
                    GridL.Columns.Add(new DataGridTextColumn { Header = column.ColumnName, Binding = new Avalonia.Data.Binding($"Row.ItemArray[{column.Ordinal}]") });
                }
            }

           
        }
    }
    private void InitializeGridR()
    {
        if (_viewModel.SelectItems != null && _viewModel.SelectItems.Tables.Count > 0)
        {
            GridR.Columns.Clear();
            GridR.ItemsSource = _viewModel.SelectItems.Tables[0].DefaultView;

            foreach (System.Data.DataColumn x in (DataContext as MainViewModel).ItemsDataSet.Tables[0].Columns)
            {
                if (x.DataType == typeof(bool))
                {
                    GridR.Columns.Add(new DataGridCheckBoxColumn { Header = x.ColumnName, Binding = new Avalonia.Data.Binding($"Row.ItemArray[{x.Ordinal}]") });
                }
                else
                {
                    GridR.Columns.Add(new DataGridTextColumn { Header = x.ColumnName, Binding = new Avalonia.Data.Binding($"Row.ItemArray[{x.Ordinal}]") });
                }
            }
        }
    }
    private void UpdateGridL()
    {
        if (_viewModel.ItemsDataSet != null && _viewModel.ItemsDataSet.Tables.Count > 0)
        {
            //GridL.BeginInit();
            GridL.ItemsSource = _viewModel.ItemsDataSet.Tables[0].DefaultView;

            /*DataView newDataView = _viewModel.ItemsDataSet.Tables[0].DefaultView;
            GridL.ItemsSource = newDataView;*/
            /*GridL.InvalidateArrange();
            GridL.InvalidateVisual();
            GridL.UpdateLayout();
            GridL.EndInit();*/

        }
    }
    private void UpdateGridR()
    {
        if (_viewModel.SelectItems != null && _viewModel.ItemsDataSet.Tables.Count > 0)
        {
            DataView newDataView = _viewModel.SelectItems.Tables[0].DefaultView;

            // Actualiser la source de données de la grille avec la nouvelle vue
            GridR.ItemsSource = newDataView;

        }
    }
}

here is my xaml code : 
<DockPanel Background="Gray">
				
				<Border BorderBrush="White" BorderThickness="0.5" CornerRadius="5">
								<StackPanel Margin="15">
									<DataGrid Background="Beige"
											  ScrollViewer.HorizontalScrollBarVisibility="Visible"
											  Height="300"
											  AutoGenerateColumns="True"
											  SelectionMode="Extended"
											  CanUserReorderColumns="True"
											  CanUserResizeColumns="True"
											  CanUserSortColumns="True"
											  IsReadOnly="True"
											  IsVisible="True"
											  x:Name="GridL"
											  ItemsSource="{Binding ItemsDataSet,Mode=OneWay}">
									</DataGrid>
									
								</StackPanel>

				</Border>
</DockPanel>
<StackPanel Grid.Column="1">
				<Button Width="50" Height="50" HorizontalAlignment="Center" HorizontalContentAlignment="Center" Content=">" Margin="3" Command="{Binding MoveItemCmd}" CommandParameter="{Binding ElementName=GridL}"  />
				<Button Width="50" Height="50" HorizontalAlignment="Center" HorizontalContentAlignment="Center" Content="&lt;" Margin="3" Command="{Binding DeleteItemCmd}" CommandParameter="{Binding ElementName=GridR}" />
</StackPanel>

<DockPanel  Grid.Column="2">
				<Border BorderBrush="White" BorderThickness="0.5" CornerRadius="5">
								<StackPanel Margin="15">
									<DataGrid Background="Beige"
											  ScrollViewer.HorizontalScrollBarVisibility="Visible"
											  Height="300"
											  AutoGenerateColumns="True"
											  SelectionMode="Extended"
											  CanUserReorderColumns="True"
											  CanUserResizeColumns="True"
											  CanUserSortColumns="True"
											  SelectedItem="{Binding SelectedRow}"
											  x:Name="GridR">
									</DataGrid>
								</StackPanel>
				</Border>
</DockPanel>

And my MainViewModel : 

public event EventHandler ItemsDataSetUpdated;
public event EventHandler SelectItemsUpdated;

private DataSet _itemsDataSet;
public DataSet ItemsDataSet
{
    get => _itemsDataSet;
    set => this.RaiseAndSetIfChanged(ref _itemsDataSet, value); 
}
private DataSet _selectItems;
public DataSet SelectItems
{
    get => _selectItems;
    set
    {
        _selectItems = value;
        this.RaisePropertyChanged(nameof(SelectItems)); // Déclenche l'événement PropertyChanged

    }
}

public ICommand MoveItemCmd { get; }
public ICommand DeleteItemCmd { get; }
public MainViewModel()
{
    ItemsDataSet = new DataSet();
    OldItemsDataSet = new DataSet();
    SelectItems = new DataSet();

    DataTable table = new();
    table.Columns.Add("Name", typeof(string));
    table.Columns.Add("Unit", typeof(string));
    table.Columns.Add("CurrValue", typeof(double));
    table.Columns.Add("Type", typeof(string));
    ItemsDataSet.Tables.Add(table);

    DataTable table1 = new();

    table1.Columns.Add("Name", typeof(string));
    table1.Columns.Add("Unit", typeof(string));
    table1.Columns.Add("CurrValue", typeof(double));
    table1.Columns.Add("Type", typeof(string));
    SelectItems.Tables.Add(table1);
    MoveItemCmd = new DataGridCmd(MoveSelectedItems);
    DeleteItemCmd = new DataGridCmd(DeleteSelectedItems);
}

private void MoveSelectedItems(DataGrid dataGrid)
{
    List<DataRowView> selectedItems = dataGrid.SelectedItems.Cast<DataRowView>().ToList();

    DataTable selectedTable = ItemsDataSet.Tables[0].Clone();

    foreach (DataRowView selectedRow in selectedItems)
    {
        selectedTable.ImportRow(selectedRow.Row);
    }

    foreach (DataRowView selectedRow in selectedItems)
    {
        ItemsDataSet.Tables[0].Rows.Remove(selectedRow.Row);

    }
    foreach (DataRow row in selectedTable.Rows)
    {
        SelectItems.Tables[0].Rows.Add(row.ItemArray);

    }
    SelectItems.AcceptChanges();
    ItemsDataSet.AcceptChanges();
    SelectItems.Tables[0].DefaultView.Sort = "Name ASC";
    ItemsDataSet.Tables[0].DefaultView.Sort = "Name ASC";
    ItemsDataSetUpdated?.Invoke(this, EventArgs.Empty);
    SelectItemsUpdated?.Invoke(this, EventArgs.Empty);

}

**Comment by timunie** [comment, marked as answer]
Use DataGridCollectionView in your VM in that case (even it breaks MVVM slightly)

**Reply by xujiaguan** [reply]
hello，how to using DataGridCollectionView to dynamic datagrid updating，I can not to find any demo.

**Reply by thevortexcloud** [reply]
There is sample code here: https://github.com/AvaloniaUI/Avalonia.Controls.DataGrid/blob/master/src/DataGridSample/DataGridPage.axaml.cs

I don't think there is any actual docs for it though.

**Comment by thevortexcloud** [comment]
Does DataSet and it's data objects use `INotifyPropertyChanged`? I can't seem to see it anywhere in the official docs.
